ray-operator/controllers/ray/batchscheduler/yunikorn/yunikorn_scheduler.go

```diff
@@ -42,19 +46,64 @@ func (y *YuniKornScheduler) DoBatchSchedulingOnSubmission(_ context.Context, _ *
 	return nil
 }
 
+// populatePodLabels is a helper function that copies RayCluster's label to the given pod based on the label key
 func (y *YuniKornScheduler) populatePodLabels(app *rayv1.RayCluster, pod *corev1.Pod, sourceKey string, targetKey string) {
 	// check labels
 	if value, exist := app.Labels[sourceKey]; exist {
-		y.log.Info("Updating pod label based on RayCluster annotations",
+		y.log.Info("Updating pod label based on RayCluster labels",
 			"sourceKey", sourceKey, "targetKey", targetKey, "value", value)
 		pod.Labels[targetKey] = value
 	}
 }
 
-func (y *YuniKornScheduler) AddMetadataToPod(app *rayv1.RayCluster, _ string, pod *corev1.Pod) {
+// AddMetadataToPod adds essential labels and annotations to the Ray pods
+// the yunikorn scheduler needs these labels and annotations in order to do the scheduling properly
+func (y *YuniKornScheduler) AddMetadataToPod(app *rayv1.RayCluster, groupName string, pod *corev1.Pod) {
+	// the applicationID and queue name must be provided in the labels
 	y.populatePodLabels(app, pod, RayClusterApplicationIDLabelName, YuniKornPodApplicationIDLabelName)
 	y.populatePodLabels(app, pod, RayClusterQueueLabelName, YuniKornPodQueueLabelName)
 	pod.Spec.SchedulerName = y.Name()
+
+	// when gang scheduling is enabled, extra annotations need to be added to all pods
+	if y.isGangSchedulingEnabled(app) {
+		// populate the taskGroups info to each pod
+		err := y.populateTaskGroupsAnnotationToPod(app, pod)
+		if err != nil {
+			y.log.Error(err, "failed to add gang scheduling related annotations to pod, "+
+				"gang scheduling will not be enabled for this workload",
+				"rayCluster", app.Name, "name", pod.Name, "namespace", pod.Namespace)
+			return
```

Comment:
Should we change the `BatchScheduler` interface to return error instead?

Reply:
I don't think we will see error [here](https://github.com/ray-project/kuberay/blob/6a064174434fa0903019922bf367cf3ab1307a2f/ray-operator/controllers/ray/batchscheduler/yunikorn/yunikorn_scheduler.go#L94-L96), its a standard json marshal, not expecting a foreign type. So logging an error seems to be sufficient to me. Please let me know if you think otherwise. Thanks

Reply:
Understood. Then how about moving this error handling into `populateTaskGroupsAnnotationToPod`? Because json marshal happens in that function and we are hard to know what errors will be thrown from that function. Maybe it will throw other errors. But inside that function we know that it is a simple json marshal error. This can improve readability.

And then move this log message to the end of the `populateTaskGroupsAnnotationToPod` function

https://github.com/ray-project/kuberay/blob/6a064174434fa0903019922bf367cf3ab1307a2f/ray-operator/controllers/ray/batchscheduler/yunikorn/yunikorn_scheduler.go#L89

so that we only log "gang scheduling is enabled" when it succeeds, and log "gang scheduling will not be enabled" when it fails.

Reply:
Understood. Then how about moving this error handling into `populateTaskGroupsAnnotationToPod`? Because json marshal happens in that function and we are hard to know what errors will be thrown from that function. Maybe it will throw other errors. But inside that function we know that it is a simple json marshal error. This can improve readability.

And then move this log message to the end of the `populateTaskGroupsAnnotationToPod` function

https://github.com/ray-project/kuberay/blob/6a064174434fa0903019922bf367cf3ab1307a2f/ray-operator/controllers/ray/batchscheduler/yunikorn/yunikorn_scheduler.go#L89

so that we only log "gang scheduling is enabled" when it succeeds, and log "gang scheduling will not be enabled" when it fails.

Reply:
Makes sense. Addressed.